# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/svg/SVGAnimatedLength.h
# Commit: 8a94c155e35c
# Full Hash: 8a94c155e35c69a524e0ecb17674bb5afa9c5ad3
# Author: Robert Longson <longsonr@gmail.com>
# Date: 2023-09-12 16:30:15
# Regressor Bug: 1287054
# File Overlap Count: 1
# Description:
#   Bug 1287054 part 1 - support rem, ch, ic and cap units for non-CSS lengths r=emilio,jgilbert
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D180618
# ==============================================================================

diff -r cd564dee524a -r 8a94c155e35c dom/svg/SVGAnimatedLength.h
--- a/dom/svg/SVGAnimatedLength.h	Mon Sep 11 23:42:37 2023 +0000
+++ b/dom/svg/SVGAnimatedLength.h	Mon Sep 11 23:50:11 2023 +0000
@@ -19,6 +19,9 @@
 #include "nsError.h"
 #include "nsMathUtils.h"
 
+struct GeckoFontMetrics;
+class nsPresContext;
+class nsFontMetrics;
 class mozAutoDocUpdate;
 class nsIFrame;
 
@@ -35,11 +38,23 @@
 
 class UserSpaceMetrics {
  public:
+  enum class Type : uint32_t { This, Root };
+  static GeckoFontMetrics DefaultFontMetrics();
+  static GeckoFontMetrics GetFontMetrics(const Element* aElement);
+  static WritingMode GetWritingMode(const Element* aElement);
+
   virtual ~UserSpaceMetrics() = default;
 
-  virtual float GetEmLength() const = 0;
-  virtual float GetExLength() const = 0;
+  virtual float GetEmLength(Type aType) const = 0;
+  float GetExLength(Type aType) const;
+  float GetChSize(Type aType) const;
+  float GetIcWidth(Type aType) const;
+  float GetCapHeight(Type aType) const;
   virtual float GetAxisLength(uint8_t aCtxType) const = 0;
+
+ protected:
+  virtual GeckoFontMetrics GetFontMetricsForType(Type aType) const = 0;
+  virtual WritingMode GetWritingModeForType(Type aType) const = 0;
 };
 
 class UserSpaceMetricsWithSize : public UserSpaceMetrics {
@@ -53,12 +68,16 @@
   explicit SVGElementMetrics(const SVGElement* aSVGElement,
                              const SVGViewportElement* aCtx = nullptr);
 
-  float GetEmLength() const override;
-  float GetExLength() const override;
+  float GetEmLength(Type aType) const override {
+    return SVGContentUtils::GetFontSize(GetElementForType(aType));
+  }
   float GetAxisLength(uint8_t aCtxType) const override;
 
  private:
   bool EnsureCtx() const;
+  const Element* GetElementForType(Type aType) const;
+  GeckoFontMetrics GetFontMetricsForType(Type aType) const override;
+  WritingMode GetWritingModeForType(Type aType) const override;
 
   const SVGElement* mSVGElement;
   mutable const SVGViewportElement* mCtx;
@@ -68,11 +87,12 @@
  public:
   explicit NonSVGFrameUserSpaceMetrics(nsIFrame* aFrame);
 
-  float GetEmLength() const override;
-  float GetExLength() const override;
+  float GetEmLength(Type aType) const override;
   gfx::Size GetSize() const override;
 
  private:
+  GeckoFontMetrics GetFontMetricsForType(Type aType) const override;
+  WritingMode GetWritingModeForType(Type aType) const override;
   nsIFrame* mFrame;
 };
 